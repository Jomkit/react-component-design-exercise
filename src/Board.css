:root {
  background-color: navy;
}

h1 {
  color: white;
}

.Board {
  display: flex;
  justify-content: center;
  margin: 0 auto;
}

.Board-fireworks {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  width: 0.5vmin;
  aspect-ratio: 1;
  animation: firework 2s infinite;
  background-size: 0.5vmin 0.5vmin;
  background-repeat: no-repeat;
}
.Board-fireworks,
.Board-fireworks::before,
.Board-fireworks::after {
  --initialSize: 0.5vmin;
  --finalSize: 45vmin;
  --particleSize: 0.2vmin;
  --color1: yellow;
  --color2: khaki;
  --color3: white;
  --color4: lime;
  --color5: gold;
  --color6: mediumseagreen;
  --y: -30vmin;
  --x: -50%;
  --initialY: 60vmin;
  content: "";
  animation: firework 2s infinite;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, var(--y));
  width: var(--initialSize);
  aspect-ratio: 1;
  background: 
    
    radial-gradient(circle, var(--color1) var(--particleSize), #0000 0) 50% 0%,
    radial-gradient(circle, var(--color2) var(--particleSize), #0000 0) 100% 50%,
    radial-gradient(circle, var(--color3) var(--particleSize), #0000 0) 50% 100%,
    radial-gradient(circle, var(--color4) var(--particleSize), #0000 0) 0% 50%,
    
    /* bottom right */
    radial-gradient(circle, var(--color5) var(--particleSize), #0000 0) 80% 90%,
    radial-gradient(circle, var(--color6) var(--particleSize), #0000 0) 95% 90%,
    radial-gradient(circle, var(--color1) var(--particleSize), #0000 0) 90% 70%,
    radial-gradient(circle, var(--color2) var(--particleSize), #0000 0) 100% 60%,
    radial-gradient(circle, var(--color3) var(--particleSize), #0000 0) 55% 80%,
    radial-gradient(circle, var(--color4) var(--particleSize), #0000 0) 70% 77%,
    
    /* bottom left */
    radial-gradient(circle, var(--color5) var(--particleSize), #0000 0) 22% 90%,
    radial-gradient(circle, var(--color6) var(--particleSize), #0000 0) 45% 90%,
    radial-gradient(circle, var(--color1) var(--particleSize), #0000 0) 33% 70%,
    radial-gradient(circle, var(--color2) var(--particleSize), #0000 0) 10% 60%,
    radial-gradient(circle, var(--color3) var(--particleSize), #0000 0) 31% 80%,
    radial-gradient(circle, var(--color4) var(--particleSize), #0000 0) 28% 77%,
    radial-gradient(circle, var(--color5) var(--particleSize), #0000 0) 13% 72%,
    
    /* top left */
    radial-gradient(circle, var(--color6) var(--particleSize), #0000 0) 80% 10%,
    radial-gradient(circle, var(--color1) var(--particleSize), #0000 0) 95% 14%,
    radial-gradient(circle, var(--color2) var(--particleSize), #0000 0) 90% 23%,
    radial-gradient(circle, var(--color3) var(--particleSize), #0000 0) 100% 43%,
    radial-gradient(circle, var(--color4) var(--particleSize), #0000 0) 85% 27%,
    radial-gradient(circle, var(--color5) var(--particleSize), #0000 0) 77% 37%,
    radial-gradient(circle, var(--color6) var(--particleSize), #0000 0) 60% 7%,
    
    /* top right */
    radial-gradient(circle, var(--color1) var(--particleSize), #0000 0) 22% 14%,
    radial-gradient(circle, var(--color1) var(--particleSize), #0000 0) 45% 20%,
    radial-gradient(circle, var(--color1) var(--particleSize), #0000 0) 33% 34%,
    radial-gradient(circle, var(--color1) var(--particleSize), #0000 0) 10% 29%,
    radial-gradient(circle, var(--color1) var(--particleSize), #0000 0) 31% 37%,
    radial-gradient(circle, var(--color1) var(--particleSize), #0000 0) 28% 7%,
    radial-gradient(circle, var(--color1) var(--particleSize), #0000 0) 13% 42%
    ;
  background-size: var(--initialSize) var(--initialSize);
  background-repeat: no-repeat;
}

.Board-fireworks::before {
  transform: translate(-50%, -50%) rotate(25deg) !important; 
}
.Board-fireworks::after {
  transform: translate(-50%, -50%) rotate(-37deg) !important;
}

@keyframes firework {
  0% { 
    transform: translate(-50%, 60vh);
    width: 0.5vmin;
    opacity: 1;
  }
  50% { 
    width: 0.5vmin;
    opacity: 1;
  }
  100% { 
    width: 45vmin; 
    opacity: 0; 
  }
}
/* .Board-win {

} */