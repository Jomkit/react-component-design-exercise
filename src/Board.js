import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *    coord zeroed at upper left (0,0)
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows = 3, ncols = 3, chanceLightStartsOn = 0.5 }) {
  const [board, setBoard] = useState(createBoard());

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard() {
    let initialBoard = [];
    // TODO: create array-of-arrays of true/false values

    // Initialize # of empty arrays according to nrows
    for(let i = 0; i < nrows; i++){
      initialBoard.push([]);
      for(let j = 0; j < ncols; j++){
        initialBoard[i].push(Math.random() < chanceLightStartsOn ? true : false);
      }
    }
    
    return initialBoard;
  }

  function hasWon() {
    // TODO: check the board in state to determine whether the player has won.
    // if every cell is false (not lit) then win
    return board.every(row => row.every(c => c == false));
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          // Clicked cell
          boardCopy[y][x] = !boardCopy[y][x];

        }
      };

      const newBoard = oldBoard.map(row => [...row]);
      flipCell(y, x, newBoard);
      // top cell if not edge [y-1][x]
      flipCell(y-1, x, newBoard);
      // right cell if not edge [y][x+1]
      flipCell(y, x+1, newBoard);
      // bottom cell if not edge [y+1][x]
      flipCell(y+1, x, newBoard);
      // left cell if not edge [y][x-1]
      flipCell(y, x-1, newBoard);
      
      return newBoard;
    });
  }

  // if the game is won, just show a winning msg & render nothing else
  { if(hasWon()) {
      return (
        <>
        <div className="Board-fireworks"></div>
        <h1 className="Board-win">You Win!</h1>
        </>
      )
    }
  }

  // TODO

  // make table board
  return (
    <div className="Board">
      <table>
        <tbody>
          {board.map((row, y) => 
            <tr key={y}>
              {row.map((c, x) => (
                <Cell 
                  key={`${y}-${x}`} 
                  flipCellsAroundMe={() => flipCellsAround(`${y}-${x}`)} 
                  isLit={c} 
                />)
              )}
            </tr>
          )}
        </tbody>

      </table>
    </div>
  )
}

export default Board;
