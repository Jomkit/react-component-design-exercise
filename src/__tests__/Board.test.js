import { render, fireEvent } from "@testing-library/react";
import Board from "../Board";

it('should match snapshot', () => {
    const { asFragment } = render(<Board nrows={5} ncols={5} chanceLightStartsOn={1.0} />);
    expect(asFragment()).toMatchSnapshot();
})

it('should match snapshot (win state)', () => {
    const { asFragment } = render(<Board nrows={5} ncols={5} chanceLightStartsOn={0.0} />);
    expect(asFragment()).toMatchSnapshot();
})

it('should handle clicks on cells', () => {
    const { container } = render(<Board nrows={5} ncols={5} chanceLightStartsOn={1.0} />);

    const cell = container.querySelector(".Cell-lit");
    expect(cell).toHaveClass("Cell-lit");

    fireEvent.click(cell);
    expect(cell).not.toHaveClass("Cell-lit");
})

it('should flip lit status of all cells around a clicked cell', () => {
    const { container, debug } = render(<Board nrows={5} ncols={5} chanceLightStartsOn={1.0} />);

    const allCells = container.querySelectorAll(".Cell");
    expect(allCells.length).toEqual(25);
    
    fireEvent.click(allCells[13]);
    
    const cellsLit = container.querySelectorAll(".Cell-lit");
    expect(cellsLit.length).toEqual(20);
})