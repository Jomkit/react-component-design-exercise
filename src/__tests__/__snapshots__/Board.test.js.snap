// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot (win state) 1`] = `
<DocumentFragment>
  <div
    class="Board-fireworks"
  />
  <h1
    class="Board-win"
  >
    You Win!
  </h1>
</DocumentFragment>
`;

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="Board"
  >
    <table>
      <tbody>
        <tr>
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
        </tr>
        <tr>
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
          <td
            class="Cell Cell-lit"
          />
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
